Set CUDA arch correctly when building with torch.utils.cpp_extension

rgommers opened the pull request:
The old behavior was to always use `sm_30`. The new behavior is:

- For building via a setup.py, check if `'arch'` is in `extra_compile_args`.  If so, don't change anything.
- If `TORCH_CUDA_ARCH_LIST` is set, respect that (can be 1 or more arches)
- Otherwise, query device capability and use that.

To test this, for example on a machine with `torch` installed for py37:
```
$ git clone https://github.com/pytorch/extension-cpp.git
$ cd extension-cpp/cuda
$ python setup.py install
$ cuobjdump --list-elf build/lib.linux-x86_64-3.7/lltm_cuda.cpython-37m-x86_64-linux-gnu.so

ELF file    1: lltm.1.sm_61.cubin
```

Existing tests in `test_cpp_extension.py` for `load_inline` and for compiling via `setup.py` in test/cpp_extensions/ cover this.

Closes gh-18657

EDIT: some more tests:

```
from torch.utils.cpp_extension import load

lltm = load(name='lltm', sources=['lltm_cuda.cpp', 'lltm_cuda_kernel.cu'])
```

```
# with TORCH_CUDA_ARCH_LIST undefined or an empty string
$ cuobjdump --list-elf /tmp/torch_extensions/lltm/lltm.so
ELF file    1: lltm.1.sm_61.cubin

# with TORCH_CUDA_ARCH_LIST = "3.5 5.2 6.0 6.1 7.0+PTX"
$ cuobjdump --list-elf build/lib.linux-x86_64-3.7/lltm_cuda.cpython-37m-x86_64-linux-gnu.so 
ELF file    1: lltm_cuda.cpython-37m-x86_64-linux-gnu.1.sm_35.cubin
ELF file    2: lltm_cuda.cpython-37m-x86_64-linux-gnu.2.sm_52.cubin
ELF file    3: lltm_cuda.cpython-37m-x86_64-linux-gnu.3.sm_60.cubin
ELF file    4: lltm_cuda.cpython-37m-x86_64-linux-gnu.4.sm_61.cubin
ELF file    5: lltm_cuda.cpython-37m-x86_64-linux-gnu.5.sm_70.cubin
```

rgommers committed:
Set CUDA arch correctly when building with torch.utils.cpp_extension

The old behavior was to always use sm_30. The new behavior is:

- For building via a setup.py, check if 'arch' is in
  `extra_compile_args`.  If so, don't change anything.
- If TORCH_CUDA_ARCH_LIST is set, respect that (can be 1 or more arches)
- Otherwise, query device capability and use that.

To test this, for example on a machine with torch installed for py37:
```
$ git clone https://github.com/pytorch/extension-cpp.git
$ cd extension-cpp/cuda
$ python setup.py install
$ cuobjdump --list-elf build/lib.linux-x86_64-3.7/lltm_cuda.cpython-37m-x86_64-linux-gnu.so

ELF file    1: lltm.1.sm_61.cubin
```

Existing tests in `test_cpp_extension.py` for `load_inline` and for
compiling via `setup.py` in test/cpp_extensions/ cover this.

Closes gh-18657

rgommers commented:
The `pytorch_linux_xenial_cuda9_cudnn7_py3_*` failures are all the same:
```
...
Jul 26 01:46:51   ERROR: Complete output from command /opt/conda/bin/python -u -c 'import setuptools, tokenize;__file__='"'"'/tmp/pip-req-build-7nnrgank/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' bdist_wheel -d /tmp/pip-wheel-0pmjh4_e --python-tag cp36:
Jul 26 01:46:51   ERROR: Building wheel torchvision-0.3.0a0+2b73a48
...
Jul 26 01:48:16       File "/opt/conda/lib/python3.6/distutils/ccompiler.py", line 574, in compile
Jul 26 01:48:16         self._compile(obj, src, ext, cc_args, extra_postargs, pp_opts)
Jul 26 01:48:16       File "/opt/conda/lib/python3.6/site-packages/torch/utils/cpp_extension.py", line 267, in unix_wrap_compile
Jul 26 01:48:16         "'-fPIC'"] + cflags + _get_cuda_arch_flags(cflags)
Jul 26 01:48:16       File "/opt/conda/lib/python3.6/site-packages/torch/utils/cpp_extension.py", line 947, in _get_cuda_arch_flags
Jul 26 01:48:16         raise ValueError("Unknown CUDA arch, {}, or GPU not supported".format(arch))
Jul 26 01:48:16     ValueError: Unknown CUDA arch, Maxwell, or GPU not supported
```
It's failing on building `torchvision` with the conda `compiler_compat`/`sysroot` thing thrown in the mix. Looking into it.

rgommers commented:
Question, how complicated do you want this code to be? I'd suggest to keep it simple and require `arch_bin` == `arch_ptx`, and implement this as simple dict lookups that gives the user one or more `-gencode=arch=compute_xx,code=sm_xx` flags. 

The alternative is to translate the code in [select_compute_arch.cmake](https://github.com/pytorch/pytorch/blob/master/cmake/Modules_CUDA_fix/upstream/FindCUDA/select_compute_arch.cmake) exactly, so it allows `compute_xx,code=sm_yy` (with `xx` != `yy`). That's probably going to be ~150 LoC.

ezyang commented:
Still reading. First thought: is there any way we can include the tests as part of the test suite? This sort of stuff is pretty subtle and so it's very helpful to have some sort of automated check when someone inevitably breaks this.

ezyang commented:
> I'd suggest to keep it simple and require arch_bin == arch_ptx, and implement this as simple dict lookups that gives the user one or more -gencode=arch=compute_xx,code=sm_xx flags.

This seems fine, if we give a reasonable error message when they stray from the happy path. If people don't like it they'll file bugs :)

ezyang commented on torch/utils/cpp_extension.py:
```diff
@@ -211,7 +211,7 @@ class BuildExtension(build_ext, object):
     When using :class:`BuildExtension`, it is allowed to supply a dictionary
     for ``extra_compile_args`` (rather than the usual list) that maps from
-    languages (``cxx`` or ``cuda``) to a list of additional compiler flags to
+    languages (``cxx`` or ``nvcc``) to a list of additional compiler flags to
```
Nice catch

ezyang commented on torch/utils/cpp_extension.py:
```diff
@@ -917,6 +919,42 @@ def _prepare_ldflags(extra_ldflags, with_cuda, verbose):
+                  '7.0': '70', '7.0+PTX': '70',  '7.2': '72', '7.2+PTX': '72',
+                  '7.5': '75', '7.5+PTX': '75',
+                  'Kepler': '52', 'Tegra': '52', 'Maxwell': '52',
+                  'Pascal': '60', 'Volta': '70', 'Turing': '75'}
```
I didn't check the correctness of this table.

rgommers replied:
for reference, here is the CMake code I based it on:
```
  list(APPEND CUDA_KNOWN_GPU_ARCHITECTURES "Kepler+Tegra" "Kepler+Tesla" "Maxwell+Tegra")
  list(APPEND CUDA_COMMON_GPU_ARCHITECTURES "5.2")

  list(APPEND CUDA_KNOWN_GPU_ARCHITECTURES "Pascal")
  list(APPEND CUDA_COMMON_GPU_ARCHITECTURES "6.0" "6.1")

  list(APPEND CUDA_KNOWN_GPU_ARCHITECTURES "Volta")
  list(APPEND CUDA_COMMON_GPU_ARCHITECTURES "7.0" "7.0+PTX")
  list(APPEND CUDA_ALL_GPU_ARCHITECTURES "7.0" "7.0+PTX" "7.2" "7.2+PTX")

  list(APPEND CUDA_KNOWN_GPU_ARCHITECTURES "Turing")
  list(APPEND CUDA_COMMON_GPU_ARCHITECTURES "7.5" "7.5+PTX")
  list(APPEND CUDA_ALL_GPU_ARCHITECTURES "7.5" "7.5+PTX")
```

Sometimes I had to choose, e.g. `Pascal` can be `6.0` or `6.1`, so I chose the lowest ones.

ezyang commented:
@pytorchbot rebase this please

rgommers committed:
Fix two flake8 issues [skip ci]

rgommers commented:
> Still reading. First thought: is there any way we can include the tests (EDIT) for this particular functionality as part of the test suite? This sort of stuff is pretty subtle and so it's very helpful to have some sort of automated check when someone inevitably breaks this.

I'll give it a try. The auto-selection is hardware-dependent, so can only be done as a self-consistency check. Setting `TORCH_CUDA_ARCH_LIST` to some values and using `cpp_extension.load` + `cuobjdump` to check the output should be feasible. I'd avoid adding multiple new `setup.py` files and invoking those (such tests tend to be more fragile than the code they test, and there's already some variations on this in CI - that's why the `Maxwell` test failed). The logic is anyway mostly common between `load`/`load_inline` and `setup.py`.

> >    I'd suggest to keep it simple and require arch_bin == arch_ptx, and implement this as simple dict lookups that gives the user one or more -gencode=arch=compute_xx,code=sm_xx flags.

> This seems fine, if we give a reasonable error message when they stray from the happy path. If people don't like it they'll file bugs :)

Any unrecognized names will trigger `ValueError("Unknown CUDA arch ({}) or GPU not supported".format(arch))`.

Note that it's always still possible for users to use
```
extra_compile_args = {'nvcc': `-gencode=arch=compute_xx,code=sm_yy`}
```
so it's not a hard limitation. I doubt many people will need this, but it's possible.

rgommers committed:
Add test for CUDA arch flags cpp_extension feature

rgommers committed:
Attempt to make Windows test for CUDA arch flags happy.

- Apparently cannot delete tmpdirs on Windows
- Extension is .pyd

rgommers commented:
Some details on the Windows failure, probably resolved by the last commit but could still be useful:

<details>

```
06:03:19 test_jit_cuda_archflags (__main__.TestCppExtension) ... Loading extension module jit_extension...
06:03:19 Not wiping extensions build folder because Windows
06:03:19 Detected CUDA files, patching ldflags
06:03:19 Emitting ninja build file C:\Windows\TEMP\tmpra465hb6\build.ninja...
06:03:19 Building extension module cudaext_archflags...
06:03:28 [1/3] cl /showIncludes -DTORCH_EXTENSION_NAME=cudaext_archflags -DTORCH_API_INCLUDE_EXTENSION_H -IC:\Jenkins\workspace\pytorch-builds\pytorch-win-ws2016-cuda9-cudnn7-py3-test2\build\win_tmp\build\torch\include -IC:\Jenkins\workspace\pytorch-builds\pytorch-win-ws2016-cuda9-cudnn7-py3-test2\build\win_tmp\build\torch\include\torch\csrc\api\include -IC:\Jenkins\workspace\pytorch-builds\pytorch-win-ws2016-cuda9-cudnn7-py3-test2\build\win_tmp\build\torch\include\TH -IC:\Jenkins\workspace\pytorch-builds\pytorch-win-ws2016-cuda9-cudnn7-py3-test2\build\win_tmp\build\torch\include\THC "-IC:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v9.0\include" -IC:\Jenkins\Miniconda3\Include -D_GLIBCXX_USE_CXX11_ABI=0 -fPIC -std=c++11 -c C:\Jenkins\workspace\pytorch-builds\pytorch-win-ws2016-cuda9-cudnn7-py3-test2\test\cpp_extensions\cuda_extension.cpp /Focuda_extension.o
```

Wrong flags but unrelated to this PR (also happens in other `cpp_extensions` tests:
```
06:03:28 cl : Command line warning D9002 : ignoring unknown option '-fPIC'
06:03:28 cl : Command line warning D9002 : ignoring unknown option '-std=c++11'
06:03:28 C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\Tools\MSVC\14.11.25503\include\xlocale(314): warning C4530: C++ exception handler used, but unwind semantics are not enabled. Specify /EHsc
06:03:28 C:\Jenkins\workspace\pytorch-builds\pytorch-win-ws2016-cuda9-cudnn7-py3-test2\build\win_tmp\build\torch\include\torch/csrc/jit/tracer.h(302): warning C4273: 'torch::jit::tracer::addInputs': inconsistent dll linkage
06:03:28 C:\Jenkins\workspace\pytorch-builds\pytorch-win-ws2016-cuda9-cudnn7-py3-test2\build\win_tmp\build\torch\include\torch/csrc/jit/tracer.h(290): note: see previous definition of 'addInputs'
06:03:28 C:\Jenkins\workspace\pytorch-builds\pytorch-win-ws2016-cuda9-cudnn7-py3-test2\build\win_tmp\build\torch\include\torch/csrc/jit/tracer.h(309): warning C4273: 'torch::jit::tracer::addInputs': inconsistent dll linkage
06:03:28 C:\Jenkins\workspace\pytorch-builds\pytorch-win-ws2016-cuda9-cudnn7-py3-test2\build\win_tmp\build\torch\include\torch/csrc/jit/tracer.h(296): note: see previous definition of 'addInputs'
06:03:28 C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\Tools\MSVC\14.11.25503\include\iosfwd(292): warning C4577: 'noexcept' used with no exception handling mode specified; termination on exception is not guaranteed. Specify /EHsc
```

Warnings are the same as gh-18425 and gh-18986 (MSVC and >= CUDA 9.2 bug).
They may not be fatal but are due to a mismatch between the installed CUDA and
the arch flags we test with (see
https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#constexpr-functions)

```
06:03:28 [1/3] cl /showIncludes -DTORCH_EXTENSION_NAME=cudaext_archflags -DTORCH_API_INCLUDE_EXTENSION_H -IC:\Jenkins\workspace\pytorch-builds\pytorch-win-ws2016-cuda9-cudnn7-py3-test2\build\win_tmp\build\torch\include -IC:\Jenkins\workspace\pytorch-builds\pytorch-win-ws2016-cuda9-cudnn7-py3-test2\build\win_tmp\build\torch\include\torch\csrc\api\include -IC:\Jenkins\workspace\pytorch-builds\pytorch-win-ws2016-cuda9-cudnn7-py3-test2\build\win_tmp\build\torch\include\TH -IC:\Jenkins\workspace\pytorch-builds\pytorch-win-ws2016-cuda9-cudnn7-py3-test2\build\win_tmp\build\torch\include\THC "-IC:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v9.0\include" -IC:\Jenkins\Miniconda3\Include -D_GLIBCXX_USE_CXX11_ABI=0 -fPIC -std=c++11 -c C:\Jenkins\workspace\pytorch-builds\pytorch-win-ws2016-cuda9-cudnn7-py3-test2\test\cpp_extensions\cuda_extension.cpp /Focuda_extension.o
...
06:03:34 [2/3] C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v9.0\bin\nvcc -DTORCH_EXTENSION_NAME=cudaext_archflags -DTORCH_API_INCLUDE_EXTENSION_H -IC:\Jenkins\workspace\pytorch-builds\pytorch-win-ws2016-cuda9-cudnn7-py3-test2\build\win_tmp\build\torch\include -IC:\Jenkins\workspace\pytorch-builds\pytorch-win-ws2016-cuda9-cudnn7-py3-test2\build\win_tmp\build\torch\include\torch\csrc\api\include -IC:\Jenkins\workspace\pytorch-builds\pytorch-win-ws2016-cuda9-cudnn7-py3-test2\build\win_tmp\build\torch\include\TH -IC:\Jenkins\workspace\pytorch-builds\pytorch-win-ws2016-cuda9-cudnn7-py3-test2\build\win_tmp\build\torch\include\THC "-IC:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v9.0\include" -IC:\Jenkins\Miniconda3\Include -D_GLIBCXX_USE_CXX11_ABI=0 -D__CUDA_NO_HALF_OPERATORS__ -D__CUDA_NO_HALF_CONVERSIONS__ -D__CUDA_NO_HALF2_OPERATORS__ --expt-relaxed-constexpr -gencode=arch=compute_52,code=sm_52 -gencode=arch=compute_61,code=sm_61 -O2 -c C:\Jenkins\workspace\pytorch-builds\pytorch-win-ws2016-cuda9-cudnn7-py3-test2\test\cpp_extensions\cuda_extension.cu -o cuda_extension.cuda.o
06:03:34 C:/Jenkins/workspace/pytorch-builds/pytorch-win-ws2016-cuda9-cudnn7-py3-test2/build/win_tmp/build/torch/include\c10/util/Exception.h(28): warning: base class dllexport/dllimport specification differs from that of the derived class
06:03:34 
06:03:34 C:/Jenkins/workspace/pytorch-builds/pytorch-win-ws2016-cuda9-cudnn7-py3-test2/build/win_tmp/build/torch/include\c10/util/Exception.h(29): warning: field of class type without a DLL interface used in a class with a DLL interface
...
06:03:34 C:/Jenkins/workspace/pytorch-builds/pytorch-win-ws2016-cuda9-cudnn7-py3-test2/build/win_tmp/build/torch/include\c10/util/typeid.h(458): warning: dllexport/dllimport conflict with "caffe2::TypeMeta::_typeMetaDataInstance [with T=caffe2::detail::_Uninitialized]"
06:03:34 (454): here; dllexport assumed
06:03:34
...
06:03:34 [3/3] "C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\Tools\MSVC\14.11.25503\bin\Hostx64\x64/link.exe" cuda_extension.o cuda_extension.cuda.o /nologo /DLL c10.lib torch.lib torch_python.lib _C.lib /LIBPATH:C:\Jenkins\Miniconda3\libs /LIBPATH:C:\Jenkins\workspace\pytorch-builds\pytorch-win-ws2016-cuda9-cudnn7-py3-test2\build\win_tmp\build\torch\lib "/LIBPATH:C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v9.0\lib/x64" cudart.lib /out:cudaext_archflags.pyd
06:03:34    Creating library cudaext_archflags.lib and object cudaext_archflags.exp
06:03:34 ERROR
```

```
06:04:50 ======================================================================
06:04:50 ERROR: test_jit_cuda_archflags (__main__.TestCppExtension)
06:04:50 ----------------------------------------------------------------------
06:04:50 Traceback (most recent call last):
06:04:50   File "test_cpp_extensions.py", line 174, in test_jit_cuda_archflags
06:04:50     self.assertEqual(p.returncode, 0)
06:04:50   File "C:\Jenkins\workspace\pytorch-builds\pytorch-win-ws2016-cuda9-cudnn7-py3-test2\test\common_utils.py", line 650, in assertEqual
06:04:50     super(TestCase, self).assertLessEqual(abs(x - y), prec, message)
06:04:50   File "C:\Jenkins\Miniconda3\lib\unittest\case.py", line 1215, in assertLessEqual
06:04:50     self.fail(self._formatMessage(msg, standardMsg))
06:04:50   File "C:\Jenkins\Miniconda3\lib\unittest\case.py", line 670, in fail
06:04:50     raise self.failureException(msg)
06:04:50 AssertionError: 1 not less than or equal to 1e-05 : 
06:04:50 
06:04:50 During handling of the above exception, another exception occurred:
06:04:50 
06:04:50 Traceback (most recent call last):
06:04:50   File "test_cpp_extensions.py", line 182, in test_jit_cuda_archflags
06:04:50     shutil.rmtree(temp_dir)
06:04:50   File "C:\Jenkins\Miniconda3\lib\shutil.py", line 494, in rmtree
06:04:50     return _rmtree_unsafe(path, onerror)
06:04:50   File "C:\Jenkins\Miniconda3\lib\shutil.py", line 389, in _rmtree_unsafe
06:04:50     onerror(os.unlink, fullname, sys.exc_info())
06:04:50   File "C:\Jenkins\Miniconda3\lib\shutil.py", line 387, in _rmtree_unsafe
06:04:50     os.unlink(fullname)
06:04:50 PermissionError: [WinError 5] Access is denied: 'C:\\Windows\\TEMP\\tmpra465hb6\\cudaext_archflags.pyd'
```

</details>

rgommers commented:
Windows tests are happy too now, I think this is good to go.

soumith commented on torch/utils/cpp_extension.py:
```diff
@@ -917,6 +919,42 @@ def _prepare_ldflags(extra_ldflags, with_cuda, verbose):
+    # First check for an env var (same as used by the main setup.py)
+    # Can be one or more architectures, e.g. '6.1' or "3.5 5.2 6.0 6.1 7.0+PTX"
+    # See cmake/Modules_CUDA_fix/upstream/FindCUDA/select_compute_arch.cmake
+    arch_list = os.environ.get('TORCH_CUDA_ARCH_LIST', None)
```
1. `TORCH_CUDA_ARCH_LIST` is commonly used with semi-colon in-between options, like here: https://github.com/pytorch/builder/blob/master/conda/pytorch-1.1.0/build.sh#L32 , so you have to also split on that. In fact, I've never seen it be split on space, I think I'm seeing it first in the comments of select_compute_arch.cmake 
2. PTX can be added to any architecture flag, not just to 7.0 / 7.2 / 7.5. For example, 3.5+PTX is totally a thing. Basically, when you add PTX, it also adds the PTX version of the flags (I have to pull up from memory how they look).
3. When something like `Pascal` is asked, then you have to include `6.0;6.1`, not just the lower one. This is important, or else there will be speed loss in the compiled kernels on particular GPUs.

rgommers replied:
Thanks for the insights @soumith.

 > In fact, I've never seen it be split on space, I think I'm seeing it first in the comments of `select_compute_arch.cmake`

Rather than accommodate both, I could just change that comment to use `;` and only allow that? Keeps the logic simpler; allowing too many combinations may be fragile. If you've never seen it, then it's likely no one is using it.

> Basically, when you add PTX, it also adds the PTX version of the flags (I have to pull up from memory how they look)

PTX is `compute_xx`. In `-gencode=arch=X,code=Y`, `X` is always `compute_##`, while `Y` can be either `sm_##`, `compute_##` (or even both). Doing all that via the env var is going to be really complicated. That's why I proposed to always use `-gencode=arch=compute_xx,code=sm_xx`, and if a user wants any other flavors she can simply pass those in via `extra_cuda_cflags` directly.

Or maybe the one extra flavor to add is `-gencode=arch=compute_xx,code=compute_xx` for `"xx+PTX"` - is that what you meant?

soumith replied:
> Rather than accommodate both, I could just change that comment to use ; and only allow that?

Yea that sounds good.

> Or maybe the one extra flavor to add is -gencode=arch=compute_xx,code=compute_xx for "xx+PTX" - is that what you meant?

Yes, I meant that the `+PTX` has to be generically parsed out for all architectures, and it seems like it's not hard to do.

rgommers replied:
All comments addressed.

> Rather than accommodate both, I could just change that comment to use ; and only allow that?

I found another example with spaces, in `docker/pytorch/Dockerfile`, so supported both delimiters after all.

soumith reviewed (changes requested):
this is an amazing and long-overdue patch. thanks for sending it.
I requested some essential changes.

rgommers committed:
Address review comments

vadimkantorov commented:
If GPU capability check is to be done at setup.py, can it also skip not very CUDA-capable devices (like small GPUs with very little memory)? https://github.com/pytorch/pytorch/issues/4716

rgommers commented:
> If GPU capability check is to be done at setup.py, can it also skip not very CUDA-capable devices (like small GPUs with very little memory)? #4716

thanks @vadimkantorov, that looks related. let me address the review comments on this PR, and then I'll look into taking gh-4716 along.

rgommers commented:
CI failures so far all look unrelated

`Python 3.7 Lint`:
```
./torch/jit/__init__.py:1446:136: B950 line too long (135 > 120 characters)
```

`binary_linux_manywheel_2.7mu_cpu_devtoolset7_test` (known issue, non-CUDA build):
```
Aug 13 00:29:58 ImportError: libmklml_intel.so: cannot open shared object file: No such file or directory
```

`pytorch_xla_linux_trusty_py3_6_gcc5_4_build` (haven't seen before, but non CUDA build as well):
```
Aug 13 01:26:18 Failed to run '['/var/lib/jenkins/workspace/xla/scripts/generate_code.sh']'
Aug 13 01:26:18 =================== sccache compilation log ===================
Aug 13 01:26:18 + cleanup
Aug 13 01:26:18 + retcode=1
Aug 13 01:26:18 + set +x
Aug 13 01:26:18 ERROR:sccache::server: Compilation failed: Output { status: ExitStatus(ExitStatus(256)), stdout: "", stderr: "/tmp/ccKvGZ7R.s: Assembler messages:\n/tmp/ccKvGZ7R.s:74: Error: operand type mismatch for `vxorps\'\n/tmp/ccKvGZ7R.s:78: Error: operand type mismatch for `vandps\'\n/tmp/ccKvGZ7R.s:114: Error: operand size mismatch for `vmovdqa64\'\n/tmp/ccKvGZ7R.s:117: Error: no such instruction: `kmovb %eax,%k3\'\n/tmp/ccKvGZ7R.s:119: Error: operand size mismatch for `vmovdqa64\'\n/tmp/ccKvGZ7R.s:120: Error: operand size mismatch for `vmovdqa64\'\n/tmp/ccKvGZ7R.s:121: Error: operand size mismatch for `vmovdqa64\'\n/tmp/ccKvGZ7R.s:122: Error: operand type mismatch for `vpxord\'\n/tmp/ccKvGZ7R.s:124: Error: no such instruction: `kmovb %eax,%k4\'\n/tmp/ccKvGZ7R.s:125: Error: operand type mismatch for `vpabsq\'\n/tmp/ccKvGZ7R.s:126: Error: operand size mismatch for `vmovdqa64\'\n" }
```

Jenkins jobs haven't started yet after 4 hours. This is good to review again I think though.

ezyang commented:
@pytorchbot rebase this please

rgommers commented:
Hmm, the `linux_xenial_cuda9_cudnn7` tests have one failure, there's a case where there should be two arches but we get only one:

```
Aug 13 15:46:24 ======================================================================
Aug 13 15:46:24 FAIL: test_jit_cuda_archflags (__main__.TestCppExtension)
Aug 13 15:46:24 ----------------------------------------------------------------------
Aug 13 15:46:24 Traceback (most recent call last):
Aug 13 15:46:24   File "test_cpp_extensions.py", line 222, in test_jit_cuda_archflags
Aug 13 15:46:24     self._run_jit_cuda_archflags(flags, expected)
Aug 13 15:46:24   File "test_cpp_extensions.py", line 185, in _run_jit_cuda_archflags
Aug 13 15:46:24     _check_cuobjdump_output(expected[0])
Aug 13 15:46:24   File "test_cpp_extensions.py", line 164, in _check_cuobjdump_output
Aug 13 15:46:24     ['sm_' + xx for xx in expected_values])
Aug 13 15:46:24   File "/var/lib/jenkins/workspace/test/common_utils.py", line 638, in assertEqual
Aug 13 15:46:24     super(TestCase, self).assertEqual(len(x), len(y), message)
Aug 13 15:46:24 AssertionError: 1 != 2 : 
```

It's either `Maxwell+Tegra;6.1` or `Pascal 3.5` that's failing, I'll need to improve the assertion message to see which of those it is.

rgommers committed:
Make CUDA arch test more robust; add better assertion messages

rgommers commented:
Still two relevant failures:

`pytorch-win-ws2016-cuda9-cudnn7-py3-test2`:
```
05:50:31 ======================================================================
05:50:31 FAIL: test_jit_cuda_archflags (__main__.TestCppExtension)
05:50:31 ----------------------------------------------------------------------
05:50:31 Traceback (most recent call last):
05:50:31   File "test_cpp_extensions.py", line 227, in test_jit_cuda_archflags
05:50:31     self._run_jit_cuda_archflags(flags, expected)
05:50:31   File "test_cpp_extensions.py", line 190, in _run_jit_cuda_archflags
05:50:31     _check_cuobjdump_output(expected[0])
05:50:31   File "test_cpp_extensions.py", line 164, in _check_cuobjdump_output
05:50:31     raise AssertionError("Flags: {},  Stderr: {}".format(flags, err))
05:50:31 AssertionError: Flags: ,  Stderr: 
05:50:31 
```

`pytorch_linux_xenial_cuda9_cudnn7_py3_test`:
```
Aug 14 06:39:25 ======================================================================
Aug 14 06:39:25 ERROR: test_jit_cuda_archflags (__main__.TestCppExtension)
Aug 14 06:39:25 ----------------------------------------------------------------------
Aug 14 06:39:25 Traceback (most recent call last):
Aug 14 06:39:25   File "/opt/conda/lib/python3.6/site-packages/torch/utils/cpp_extension.py", line 1027, in _build_extension_module
Aug 14 06:39:25     check=True)
Aug 14 06:39:25   File "/opt/conda/lib/python3.6/subprocess.py", line 438, in run
Aug 14 06:39:25     output=stdout, stderr=stderr)
Aug 14 06:39:25 subprocess.CalledProcessError: Command '['ninja', '-v']' returned non-zero exit status 1.
Aug 14 06:39:25 
Aug 14 06:39:25 During handling of the above exception, another exception occurred:
Aug 14 06:39:25 
Aug 14 06:39:25 Traceback (most recent call last):
Aug 14 06:39:25   File "test_cpp_extensions.py", line 227, in test_jit_cuda_archflags
Aug 14 06:39:25     self._run_jit_cuda_archflags(flags, expected)
Aug 14 06:39:25   File "test_cpp_extensions.py", line 184, in _run_jit_cuda_archflags
Aug 14 06:39:25     build_directory=temp_dir,
Aug 14 06:39:25   File "/opt/conda/lib/python3.6/site-packages/torch/utils/cpp_extension.py", line 660, in load
Aug 14 06:39:25     is_python_module)
Aug 14 06:39:25   File "/opt/conda/lib/python3.6/site-packages/torch/utils/cpp_extension.py", line 829, in _jit_compile
Aug 14 06:39:25     with_cuda=with_cuda)
Aug 14 06:39:25   File "/opt/conda/lib/python3.6/site-packages/torch/utils/cpp_extension.py", line 882, in _write_ninja_file_and_build
Aug 14 06:39:25     _build_extension_module(name, build_directory, verbose)
Aug 14 06:39:25   File "/opt/conda/lib/python3.6/site-packages/torch/utils/cpp_extension.py", line 1040, in _build_extension_module
Aug 14 06:39:25     raise RuntimeError(message)
Aug 14 06:39:25 RuntimeError: Error building extension 'cudaext_archflags_v3'
```

rgommers committed:
Fix failures on CUDA 9, make assertion messages even more verbose

rgommers commented:
The `'torch::jit::tracer::addInputs': inconsistent dll linkage` issue (expand `<details>` above to see) on `pytorch-win-ws2016-cuda9-cudnn7-py3-test2` is the same as in gh-21706. There seemed to be multiple issues there with a particular MSVC version, and while the `swap` issue was fixed, the `addInputs` one wasn't. 

Somehow MSVC seems not happy with multiple redefinitions of `addInputs` in combination with templating (`torch/csrc/jit/tracer.h(290): note: see previous definition of 'addInputs'`).

@peterjc123 you looked at this before in gh-21706, do you have any advice?

rgommers commented:
Looked a bit more carefully, `std::vector` and `std::unordered_map` do indeed seem to be redefined. The former on like 266 and then 290, the latter on line 296 and 306.

@ezyang not sure I should be trying to fix that here (mixed two issues, and I don't have a Windows box so will be testing via CI), or skip the test on Windows for now and open a new issue?

rgommers commented:
CI finished:

- the three `pytorch_linux` failures are `test_conv_api` and `test_linear_api` (unrelated)
- the Lint failures are in `test_quantized.py` (unrelated)
- `caffe2-py2-cuda9.0-cudnn7-windows-build` seems to be a known issue (see pytorch-ci-hud)
- the `pytorch-win-ws2016-cuda9-cudnn7-py3` is the only failure due to this PR, as commented on above

peterjc123 commented:
@rgommers What are the actual errors? Warnings should not break the build.

peterjc123 commented on test/test_cpp_extensions.py:
```diff
@@ -141,6 +145,93 @@ def test_jit_cuda_extension(self):
+        def _check_cuobjdump_output(expected_values, is_ptx=False):
+            elf_or_ptx = '--list-ptx' if is_ptx else '--list-elf'
+            # Note, .extension name may include _v1, _v2, so first find exact name
+            ext_filename = glob.glob(os.path.join(temp_dir, 'cudaext_archflag*'))[0]
```
Are you sure this one selected the expected file? I ask this because I saw this line `Output: cuobjdump info    : File 'C:\Windows\TEMP\tmpn5dwlrij\cudaext_archflags.exp' does not contain device code` in the log. But in Windows, `.exp` file is only a file that contains the symbol table, the object file has the suffix `.obj`.

rgommers replied:
ah, staring for too long at similar build log - apologies for the noise. I'll put back the bit that selects the `.pyd` file only on Windows.

rgommers commented:
@leminhhiep20691 can you please stop approving PRs on this repo? you've just approved 13 PRs, some many times - that is just spam. thank you

rgommers committed:
Fix Windows test failure.

On Linux one just gets a `.so` file with the name
`cudaext_archflags` or `cudaext_archflags_vX`.  On Windows, there's
not only a `.pyd` file but also a `.exp`.

peterjc123 commented:
@rgommers Looks like this fixes the error, now it stops with a related error in the quantization module.

rgommers commented:
Okay, this should be good to go now - all CI failures are unrelated.

soumith commented on torch/utils/cpp_extension.py:
```diff
@@ -928,6 +930,71 @@ def _prepare_ldflags(extra_ldflags, with_cuda, verbose):
+            if 'arch' in flag:
+                return []
+
+    named_arches = {'Fermi': '2.0;2.1',
```
we can remove Fermi

soumith commented on torch/utils/cpp_extension.py:
```diff
@@ -928,6 +930,71 @@ def _prepare_ldflags(extra_ldflags, with_cuda, verbose):
+                    'Volta': '7.0+PTX',
+                    'Turing': '7.5+PTX'}
+
+    supported_arches = ['2.0', '2.1', '3.0', '3.2', '3.5', '3.7', '5.0', '5.2',
```
remove 2.0, 2.1, 3.0, 3.2

soumith commented on torch/utils/cpp_extension.py:
```diff
@@ -928,6 +930,71 @@ def _prepare_ldflags(extra_ldflags, with_cuda, verbose):
+                return []
+
+    named_arches = {'Fermi': '2.0;2.1',
+                    'Kepler+Tegra': '3.2',
```
remove Kepler+Tegra, we dont support it.

soumith replied:
(compilation will actually fail)

soumith commented on torch/utils/cpp_extension.py:
```diff
@@ -928,6 +930,71 @@ def _prepare_ldflags(extra_ldflags, with_cuda, verbose):
+    named_arches = {'Fermi': '2.0;2.1',
+                    'Kepler+Tegra': '3.2',
+                    'Kepler+Tesla': '3.7',
+                    'Kepler': '3.0;3.5+PTX',
```
remove 3.0 from Kepler, compilation will actually fail

soumith reviewed (changes requested):
requested small changes for stuff we dont support. I know it deviates from cmake, but cmake is outdated in this respect.

let me know if you want to make the changes, or rather keep cmake in sync with this. I'm okay with this PR landing as-is.

rgommers commented:
> I know it deviates from cmake, but cmake is outdated in this respect. let me know if you want to make the changes, or rather keep cmake in sync with this. 

If you want to update CMake too, I'd be happy to send a follow-up PR removing it from both here and CMake. If you don't want to touch CMake (`FindCUDA.cmake` and `select_compute_arch.cmake`), then I'll push another commit here.

soumith reviewed (approved):
it makes sense to update cmake too. let's just figure it out in a follow-up PR

soumith commented:
@pytorchbot rebase this please

Commit:
Merge remote-tracking branch 'origin/master' into HEAD

soumith commented:
waiting for rebased CI to pass, wanted to make sure the Windows Jenkins passes

rgommers commented:
> waiting for rebased CI to pass, wanted to make sure the Windows Jenkins passes

`pytorch-win-ws2016-cuda9-cudnn7-py3` passed, the `caffe2-py2-cuda9.0-cudnn7-windows-build` failure is identical to the failures on other PRs.

ezyang commented:
On master this is failing for some reason:

```
Aug 16 14:07:44 ======================================================================
Aug 16 14:07:44 ERROR: test_jit_cuda_archflags (__main__.TestCppExtension)
Aug 16 14:07:44 ----------------------------------------------------------------------
Aug 16 14:07:44 Traceback (most recent call last):
Aug 16 14:07:44   File "test_cpp_extensions.py", line 235, in test_jit_cuda_archflags
Aug 16 14:07:44     self._run_jit_cuda_archflags(flags, expected)
Aug 16 14:07:44   File "test_cpp_extensions.py", line 190, in _run_jit_cuda_archflags
Aug 16 14:07:44     build_directory=temp_dir,
Aug 16 14:07:44   File "/opt/conda/lib/python2.7/site-packages/torch/utils/cpp_extension.py", line 660, in load
Aug 16 14:07:44     is_python_module)
Aug 16 14:07:44   File "/opt/conda/lib/python2.7/site-packages/torch/utils/cpp_extension.py", line 829, in _jit_compile
Aug 16 14:07:44     with_cuda=with_cuda)
Aug 16 14:07:44   File "/opt/conda/lib/python2.7/site-packages/torch/utils/cpp_extension.py", line 878, in _write_ninja_file_and_build
Aug 16 14:07:44     with_cuda=with_cuda)
Aug 16 14:07:44   File "/opt/conda/lib/python2.7/site-packages/torch/utils/cpp_extension.py", line 1111, in _write_ninja_file
Aug 16 14:07:44     cuda_flags = common_cflags + COMMON_NVCC_FLAGS + _get_cuda_arch_flags()
Aug 16 14:07:44   File "/opt/conda/lib/python2.7/site-packages/torch/utils/cpp_extension.py", line 988, in _get_cuda_arch_flags
Aug 16 14:07:44     raise ValueError("Unknown CUDA arch ({}) or GPU not supported".format(arch))
Aug 16 14:07:44 ValueError: Unknown CUDA arch (5.2+PTX+Tegra) or GPU not supported
Aug 16 14:07:44 
Aug 16 14:07:44 ----------------------------------------------------------------------
```

ezyang commented:
https://circleci.com/gh/pytorch/pytorch/2485792?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link/console

rgommers commented:
Hmm, looking into it now

rgommers commented:
Okay, I think what's going on here is that `Maxwell+Tegra` is parsed out through `named_arches` into  `5.2+PTX+Tegra`. This is a Python 2.7 failure; CI passed on this PR but fails on master likely because on py27 the order of the `named_arches` dict isn't fixed, and the string replacement is picking up only `Maxwell` first.

Should use `collections.OrderedDict` here. Sending a patch asap.

rgommers commented:
CI failure should be fixed by gh-24501
